lang/java/reef-runtime-mock/src/test/java/org/apache/reef/mock/MockApplication.java

```diff
@@ -0,0 +1,266 @@
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * mock application.
+ */
+@Unit
+public class MockApplication {
+
+  private static final Logger LOG = Logger.getLogger(MockApplication.class.getName());
+
+  private final EvaluatorRequestor evaluatorRequestor;
+
+  private final Clock clock;
+
+  private final Map<String, Map<String, ActiveContext>> evaluatorId2ContextId2ContextMap = new HashMap<>();
+
+  private final Map<String, AllocatedEvaluator> evaluatorMap = new HashMap<>();
+
+  private final Map<String, FailedEvaluator> failedEvaluatorMap = new HashMap<>();
+
+  private final Map<String, RunningTask> evaluatorIdRunningTaskMap = new HashMap<>();
+
+  private final Set<FailedContext> failedContextSet = new HashSet<>();
+
+  private final Set<FailedTask> failedTaskSet = new HashSet<>();
+
+  private final Set<SuspendedTask> suspendedTaskSet = new HashSet<>();
+
+  private boolean running = false;
+
+  @Inject
+  MockApplication(final EvaluatorRequestor evaluatorRequestor, final Clock clock) {
+    this.evaluatorRequestor = evaluatorRequestor;
+    this.clock = clock;
+  }
+
+  ActiveContext getContext(final AllocatedEvaluator evaluator, final String identifier) {
+    return this.evaluatorId2ContextId2ContextMap.get(evaluator.getId()).get(identifier);
+  }
+
+  Collection<RunningTask> getRunningTasks() {
+    return this.evaluatorIdRunningTaskMap.values();
+  }
+
+  Collection<AllocatedEvaluator> getAllocatedEvaluators() {
+    return this.evaluatorMap.values();
+  }
+
+  Collection<FailedEvaluator> getFailedEvaluators() {
+    return this.failedEvaluatorMap.values();
+  }
+
+  Collection<FailedTask> getFailedTasks() {
+    return this.failedTaskSet;
+  }
+
+  Collection<FailedContext> getFailedContext() {
+    return this.failedContextSet;
+  }
+
+  void requestEvaluators(final int numEvaluators) {
+    LOG.log(Level.INFO, "request {0} Evaluators", numEvaluators);
+    evaluatorRequestor.newRequest()
+        .setMemory(128)
+        .setNumberOfCores(1)
+        .setNumber(numEvaluators)
+        .submit();
+  }
+
+  void submitTask(final ActiveContext context, final String identifier) {
+    context.submitTask(TaskConfiguration.CONF
+        .set(TaskConfiguration.IDENTIFIER, identifier)
+        .set(TaskConfiguration.TASK, DummyTestTask.class)
+        .build());
+  }
+
+  void submitContext(final ActiveContext context, final String identifier) {
+    context.submitContext(ContextConfiguration.CONF
+        .set(ContextConfiguration.IDENTIFIER, identifier)
+        .build());
+  }
+
+  boolean isRunning() {
+    return this.running;
+  }
+
+  boolean exists(final AllocatedEvaluator evaluator) {
+    return this.evaluatorMap.containsKey(evaluator.getId());
+  }
+
+  /**
+   * Job Driver is ready and the clock is set up: request the evaluatorMap.
+   */
+  final class StartHandler implements EventHandler<StartTime> {
+    @Override
+    public void onNext(final StartTime startTime) {
+      running = true;
```

Comment:
Shouldn't this ask for Evaluators?

Reply:
What you raise is a valid concern. The purpose of the MockApplication is to serve internal mock unit tests. However, if the Mock Runtime were truly mimicking REEF, then a shutdown would occur unless an alarm is set. 

The current mock runtime does not support idle semantics of REEF. It will take a little more work to do this. I can either:
1. do it in this current PR.
2. do it in another/subsequent PR. 

What say you? @markusweimer

Reply:
Follow up. I implemented the solution, which was not too bad. My next checkin will mimic REEF idle shutdown semantics.